firedrake/slate/slate.py

```diff
@@ -0,0 +1,673 @@
+algebra compiler, which interprets SLATE expressions and
+produces C++ kernel functions to be executed within the
+Firedrake architecture.
+"""
+
+import operator
+
+from ufl.algorithms.map_integrands import map_integrand_dags
+from ufl.algorithms.multifunction import MultiFunction
+from ufl.domain import join_domains, sort_domains
+
+
+__all__ = ['Tensor', 'Scalar', 'Vector', 'Matrix',
+           'Inverse', 'Transpose', 'UnaryOp', 'Negative', 'Positive',
+           'BinaryOp', 'TensorAdd', 'TensorSub', 'TensorMul']
+
+
+class CheckRestrictions(MultiFunction):
+    """UFL MultiFunction for enforcing cell-wise integrals to contain
+    only positive restrictions."""
+
+    expr = MultiFunction.reuse_if_untouched
+
+    def negative_restrictions(self, o):
+        raise ValueError("Cell-wise integrals must contain only positive restrictions.")
+
+
+class Tensor(object):
+    """An abstract representation of a finite element tensor in SLATE.
+    All derived tensors, `Scalar`, `Vector`, and `Matrix`, as well as all
+    `BinaryOp` and `UnaryOp` objects derive from this abstract base class.
+    """
+    # Initializing tensor id class variable for output purposes
+    id = 0
+
+    def __init__(self):
+        """Constructor for the Tensor class."""
+        self.id = Tensor.id
+        Tensor.id += 1
+
+        # Compute Tensor shape information
+        shape = []
+        shapes = {}
+
+        for i, arg in enumerate(self.arguments()):
+            V = arg.function_space()
+            shapelist = []
+
+            for fs in V:
+                shapelist.append(fs.fiat_element.space_dimension() * fs.dim)
+
+            shapes[i] = tuple(shapelist)
+            shape.append(sum(shapelist))
+
+        self.shapes = shapes
+        self.shape = tuple(shape)
+        self.rank = len(self.arguments())
+
+        self._integral_domains = None
+        self._subdomain_data = None
+        self._hash = None
+
+    def __add__(self, other):
+        return TensorAdd(self, other)
+
+    def __radd__(self, other):
+        """Ordering of tensor addition does not matter."""
+        return self.__add__(other)
+
+    def __sub__(self, other):
+        return TensorSub(self, other)
+
+    def __rsub__(self, other):
+        """Ordering of tensor subtraction does not matter."""
+        return self.__sub__(other)
+
+    def __mul__(self, other):
+        return TensorMul(self, other)
+
+    def __rmul__(self, other):
+        """Tensor multiplication is not commutative in general."""
+        return TensorMul(other, self)
+
+    def __neg__(self):
+        return Negative(self)
+
+    def __pos__(self):
+        return Positive(self)
+
+    @property
+    def inv(self):
+        return Inverse(self)
+
+    @property
+    def T(self):
+        return Transpose(self)
+
+    @property
+    def operands(self):
+        """Returns the tensor objects which this tensor operates on."""
+        return ()
+
+    @classmethod
+    def check_integrals(cls, integrals):
```

Comment:
You don't use `cls` in this function, so this could be a `staticmethod`. Or maybe even a free function.